internal/appsec/emitter/httpsec/http.go

```diff
@@ -108,61 +108,74 @@ func makeCookies(parsed []*http.Cookie) map[string][]string {
 	return cookies
 }
 
-// WrapHandler wraps the given HTTP handler with the abstract HTTP operation defined by HandlerOperationArgs and
-// HandlerOperationRes.
-// The onBlock params are used to cleanup the context when needed.
-// It is a specific patch meant for Gin, for which we must abort the
-// context since it uses a queue of handlers and it's the only way to make
-// sure other queued handlers don't get executed.
-// TODO: this patch must be removed/improved when we rework our actions/operations system
-func WrapHandler(handler http.Handler, span ddtrace.Span, pathParams map[string]string, opts *Config) http.Handler {
+func BeforeHandle(w http.ResponseWriter, r *http.Request, span ddtrace.Span, pathParams map[string]string, opts *Config) (
+	http.ResponseWriter, *http.Request, func(), bool) {
 	if opts == nil {
 		opts = defaultWrapHandlerConfig
 	} else if opts.ResponseHeaderCopier == nil {
 		opts.ResponseHeaderCopier = defaultWrapHandlerConfig.ResponseHeaderCopier
 	}
 
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		op, blockAtomic, ctx := StartOperation(r.Context(), HandlerOperationArgs{
-			Method:      r.Method,
-			RequestURI:  r.RequestURI,
-			Host:        r.Host,
-			RemoteAddr:  r.RemoteAddr,
-			Headers:     r.Header,
-			Cookies:     makeCookies(r.Cookies()),
-			QueryParams: r.URL.Query(),
-			PathParams:  pathParams,
-		})
-		r = r.WithContext(ctx)
-
-		defer func() {
-			var statusCode int
-			if res, ok := w.(interface{ Status() int }); ok {
-				statusCode = res.Status()
+	op, blockAtomic, ctx := StartOperation(r.Context(), HandlerOperationArgs{
+		Method:      r.Method,
+		RequestURI:  r.RequestURI,
+		Host:        r.Host,
+		RemoteAddr:  r.RemoteAddr,
+		Headers:     r.Header,
+		Cookies:     makeCookies(r.Cookies()),
+		QueryParams: r.URL.Query(),
+		PathParams:  pathParams,
+	})
+	tr := r.WithContext(ctx)
+
+	afterHandle := func() {
+		var statusCode int
+		if res, ok := w.(interface{ Status() int }); ok {
+			statusCode = res.Status()
+		}
```

Comment:
Consider adding an `else` that prints a debug statement indicating the writer somehow misses the `Status()` method (+ mention its actual type).

```suggestion
		if res, ok := w.(interface{ Status() int }); ok {
			statusCode = res.Status()
		} else {
			log.Debugf("http.ResponseWriter (%T) lacks Statsus()int method", w)
		}
```